config/utils.go

```diff
@@ -0,0 +1,133 @@
+func desanitize(actual, target *model.Config) {
+	if target.LdapSettings.BindPassword != nil && *target.LdapSettings.BindPassword == model.FAKE_SETTING {
+		*target.LdapSettings.BindPassword = *actual.LdapSettings.BindPassword
+	}
+
+	if *target.FileSettings.PublicLinkSalt == model.FAKE_SETTING {
+		*target.FileSettings.PublicLinkSalt = *actual.FileSettings.PublicLinkSalt
+	}
+	if *target.FileSettings.AmazonS3SecretAccessKey == model.FAKE_SETTING {
+		target.FileSettings.AmazonS3SecretAccessKey = actual.FileSettings.AmazonS3SecretAccessKey
+	}
+
+	if *target.EmailSettings.InviteSalt == model.FAKE_SETTING {
+		target.EmailSettings.InviteSalt = actual.EmailSettings.InviteSalt
+	}
+	if *target.EmailSettings.SMTPPassword == model.FAKE_SETTING {
+		target.EmailSettings.SMTPPassword = actual.EmailSettings.SMTPPassword
+	}
+
+	if *target.GitLabSettings.Secret == model.FAKE_SETTING {
+		target.GitLabSettings.Secret = actual.GitLabSettings.Secret
+	}
+
+	if *target.SqlSettings.DataSource == model.FAKE_SETTING {
+		*target.SqlSettings.DataSource = *actual.SqlSettings.DataSource
+	}
+	if *target.SqlSettings.AtRestEncryptKey == model.FAKE_SETTING {
+		target.SqlSettings.AtRestEncryptKey = actual.SqlSettings.AtRestEncryptKey
+	}
+
+	if *target.ElasticsearchSettings.Password == model.FAKE_SETTING {
+		*target.ElasticsearchSettings.Password = *actual.ElasticsearchSettings.Password
+	}
+
+	target.SqlSettings.DataSourceReplicas = make([]string, len(actual.SqlSettings.DataSourceReplicas))
+	for i := range target.SqlSettings.DataSourceReplicas {
+		target.SqlSettings.DataSourceReplicas[i] = actual.SqlSettings.DataSourceReplicas[i]
+	}
+
+	target.SqlSettings.DataSourceSearchReplicas = make([]string, len(actual.SqlSettings.DataSourceReplicas))
+	for i := range target.SqlSettings.DataSourceSearchReplicas {
+		target.SqlSettings.DataSourceSearchReplicas[i] = actual.SqlSettings.DataSourceSearchReplicas[i]
+	}
+}
+
+// fixConfig patches invalid or missing data in the configuration, returning true if changed.
+func fixConfig(cfg *model.Config) bool {
+	changed := false
+
+	// Ensure SiteURL has no trailing slash.
+	if strings.HasSuffix(*cfg.ServiceSettings.SiteURL, "/") {
+		*cfg.ServiceSettings.SiteURL = strings.TrimRight(*cfg.ServiceSettings.SiteURL, "/")
+		changed = true
+	}
+
+	// Ensure the directory for a local file store has a trailing slash.
+	if *cfg.FileSettings.DriverName == model.IMAGE_DRIVER_LOCAL {
+		if !strings.HasSuffix(*cfg.FileSettings.Directory, "/") {
+			*cfg.FileSettings.Directory += "/"
+			changed = true
+		}
+	}
+
+	if FixInvalidLocales(cfg) {
+		changed = true
+	}
+
+	return changed
+}
+
+// FixInvalidLocales checks and corrects the given config for invalid locale-related settings.
+//
+// Ideally, this function would be completely internal, but it's currently exposed to allow the cli
+// to test the config change before allowing the save.
+func FixInvalidLocales(cfg *model.Config) bool {
```

Comment:
I changed the signature here to make it easier to use with `fixConfig` and to avoid having weird error propagation semantics, but the gist of the original method remains untouched.